### ISSUE TITLE
Enable Websocket port with nginx

### ISSUE THREAD

**Issue by paschaldev** [dokku/dokku#3715]
## Description of problem
I have a `laravel/lumen` app that works fine. Now I need to add websocket feature which works flawlessly on localhost. The socket server is started with `php artisan websocket:serve` (This is already added to the Procfile) and it runs on port `6001`. I can't seem to find a way to expose this port publicly.

### How reproducible
This is my `nginx.conf` file
```
server {
  listen      [::]:6001;
  listen      6001;
  server_name api.apppppp.com;
  access_log  /var/log/nginx/api-access.log;
  error_log   /var/log/nginx/api-error.log;

  location    / {

    proxy_pass  http://api-6001;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
    proxy_set_header Host $http_host;
  }

server {
  listen      [::]:80;
  listen      80;
  server_name api.apppppp.com; 
  access_log  /var/log/nginx/api-access.log;
  error_log   /var/log/nginx/api-error.log;

  return 301 https://$host:443$request_uri;

}

server {
  listen      [::]:443 ssl http2;
  listen      443 ssl http2;
  
  server_name api.apppppp.com; 
  access_log  /var/log/nginx/api-access.log;
  error_log   /var/log/nginx/api-error.log;

  ssl_certificate           /home/dokku/api/tls/server.crt;
  ssl_certificate_key       /home/dokku/api/tls/server.key;
  ssl_protocols             TLSv1.2 TLSv1.3;
  ssl_prefer_server_ciphers off;

  keepalive_timeout   70;
  

  location    / {

    gzip on;
    gzip_min_length  1100;
    gzip_buffers  4 32k;
    gzip_types    text/css text/javascript text/xml text/plain text/x-component application/javascript application/x-javascript application/json application/xml  application/rss+xml font/truetype application/x-font-ttf font/opentype application/vnd.ms-fontobject image/svg+xml;
    gzip_vary on;
    gzip_comp_level  6;

    proxy_pass  http://api-5000;
    http2_push_preload on; 
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $http_connection;
    proxy_set_header Host $http_host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Port $server_port;
    proxy_set_header X-Request-Start $msec;
  }
  include /home/dokku/api/nginx.conf.d/*.conf;

  error_page 400 401 402 403 405 406 407 408 409 410 411 412 413 414 415 416 417 418 420 422 423 424 426 428 429 431 444 449 450 451 /400-error.html;
  location /400-error.html {
    root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
    internal;
  }

  error_page 404 /404-error.html;
  location /404-error.html {
    root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
    internal;
  }

  error_page 500 501 503 504 505 506 507 508 509 510 511 /500-error.html;
  location /500-error.html {
    root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
    internal;
  }

  error_page 502 /502-error.html;
  location /502-error.html {
    root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
    internal;
  }
}

upstream api-6001 {

  server 172.17.0.12:6001;
}

upstream api-5000 {

  server 172.17.0.12:5000;
}
```

### Steps to Reproduce

1. Added websocket configuration to nginx config file
2. Proxied port `6001` with `dokku proxy:ports-add api http:6001:6001`

#### Actual Results

Scanning shows that port `6001` is not open

#### Expected Results

Able to connect to internal application websocket server

### `dokku report APP_NAME` output

```
-----> uname: Linux ip-172-31-16-211 4.15.0-1039-aws #41-Ubuntu SMP Wed May 8 10:43:54 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
-----> memory: 
                     total        used        free      shared  buff/cache   available
       Mem:            983         499          72          68         412         272
       Swap:          1999         570        1429
-----> docker version: 
       Client: Docker Engine - Community
        Version:           19.03.3
        API version:       1.40
        Go version:        go1.12.10
        Git commit:        a872fc2f86
        Built:             Tue Oct  8 00:59:59 2019
        OS/Arch:           linux/amd64
        Experimental:      false
       
       Server: Docker Engine - Community
        Engine:
         Version:          19.03.3
         API version:      1.40 (minimum version 1.12)
         Go version:       go1.12.10
         Git commit:       a872fc2f86
         Built:            Tue Oct  8 00:58:31 2019
         OS/Arch:          linux/amd64
         Experimental:     false
        containerd:
         Version:          1.2.10
         GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
        runc:
         Version:          1.0.0-rc8+dev
         GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
        docker-init:
         Version:          0.18.0
         GitCommit:        fec3683
-----> docker daemon info: 
WARNING: No swap limit support
       Client:
        Debug Mode: true
       
       Server:
        Containers: 23
         Running: 23
         Paused: 0
         Stopped: 0
        Images: 61
        Server Version: 19.03.3
        Storage Driver: overlay2
         Backing Filesystem: extfs
         Supports d_type: true
         Native Overlay Diff: true
        Logging Driver: json-file
        Cgroup Driver: cgroupfs
        Plugins:
         Volume: local
         Network: bridge host ipvlan macvlan null overlay
         Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
        Swarm: inactive
        Runtimes: runc
        Default Runtime: runc
        Init Binary: docker-init
        containerd version: b34a5c8af56e510852c35414db4c1f4fa6172339
        runc version: 3e425f80a8c931f88e6d94a8c831b9d5aa481657
        init version: fec3683
        Security Options:
         apparmor
         seccomp
          Profile: default
        Kernel Version: 4.15.0-1039-aws
        Operating System: Ubuntu 18.04.3 LTS
        OSType: linux
        Architecture: x86_64
        CPUs: 1
        Total Memory: 983.9MiB
        Name: ip-172-31-16-211
        ID: 7PP3:NSGE:6GPB:TBZZ:ZOID:Q7I5:ITAV:PGUB:AJ2P:CO5A:3ZQJ:HXUZ
        Docker Root Dir: /var/lib/docker
        Debug Mode: false
        Registry: https://index.docker.io/v1/
        Labels:
        Experimental: false
        Insecure Registries:
         127.0.0.0/8
        Live Restore Enabled: false
       
-----> sigil version: 0.4.0
-----> herokuish version: 
       herokuish: 0.5.3
       buildpacks:
         heroku-buildpack-multi     v1.0.0
         heroku-buildpack-ruby      v201
         heroku-buildpack-nodejs    v161
         heroku-buildpack-clojure   v84
         heroku-buildpack-python    v154
         heroku-buildpack-java      v63
         heroku-buildpack-gradle    v30
         heroku-buildpack-scala     v85
         heroku-buildpack-play      v26
         heroku-buildpack-php       v159
         heroku-buildpack-go        v126
         buildpack-nginx            v12
-----> dokku version: dokku version 0.19.1
-----> dokku plugins: 
       plugn: 0.3.0
         00_dokku-standard    0.19.1 enabled    dokku core standard plugin
         20_events            0.19.1 enabled    dokku core events logging plugin
         app-json             0.19.1 enabled    dokku core app-json plugin
         apps                 0.19.1 enabled    dokku core apps plugin
         builder-dockerfile   0.19.1 enabled    dokku core builder-dockerfile plugin
         builder-herokuish    0.19.1 enabled    dokku core builder-herokuish plugin
         buildpacks           0.19.1 enabled    dokku core buildpacks plugin
         certs                0.19.1 enabled    dokku core certificate management plugin
         checks               0.19.1 enabled    dokku core checks plugin
         common               0.19.1 enabled    dokku core common plugin
         config               0.19.1 enabled    dokku core config plugin
         docker-options       0.19.1 enabled    dokku core docker-options plugin
         domains              0.19.1 enabled    dokku core domains plugin
         enter                0.19.1 enabled    dokku core enter plugin
         git                  0.19.1 enabled    dokku core git plugin
         letsencrypt          0.9.1 enabled    Automated installation of let's encrypt TLS certificates
         logs                 0.19.1 enabled    dokku core logs plugin
         network              0.19.1 enabled    dokku core network plugin
         nginx-vhosts         0.19.1 enabled    dokku core nginx-vhosts plugin
         plugin               0.19.1 enabled    dokku core plugin plugin
         proxy                0.19.1 enabled    dokku core proxy plugin
         ps                   0.19.1 enabled    dokku core ps plugin
         repo                 0.19.1 enabled    dokku core repo plugin
         resource             0.19.1 enabled    dokku core resource plugin
         scheduler-docker-local 0.19.1 enabled    dokku core scheduler-docker-local plugin
         shell                0.19.1 enabled    dokku core shell plugin
         ssh-keys             0.19.1 enabled    dokku core ssh-keys plugin
         storage              0.19.1 enabled    dokku core storage plugin
         tags                 0.19.1 enabled    dokku core tags plugin
         tar                  0.19.1 enabled    dokku core tar plugin
         trace                0.19.1 enabled    dokku core trace plugin
=====> api app information
       App dir:                       /home/dokku/api          
       Git sha:                       9085b29                  
       Deploy source:                 git                      
       Locked:                        false                    
=====> api buildpacks information
       Buildpacks list:               
OU = Domain Control Validated, OU = PositiveSSL Wildcard, CN = *.mhoori.com
error 20 at 0 depth lookup: unable to get local issuer certificate
=====> api ssl information
       Ssl dir:                       /home/dokku/api/tls      
       Ssl enabled:                   true                     
       Ssl hostnames:                 *.mhoori.com mhoori.com  
       Ssl expires at:                May 27 23:59:59 2020 GMT 
       Ssl issuer:                    C = GB, ST = Greater Manchester, L = Salford, O = Sectigo Limited, CN = Sectigo RSA Domain Validation Secure Server CA
       Ssl starts at:                 May 28 00:00:00 2019 GMT 
       Ssl subject:                   subject=OU = Domain Control Validated, OU = PositiveSSL Wildcard, CN = *.mhoori.com
       Ssl verified:                  self signed              
=====> api checks information
       Checks disabled list:          none                     
       Checks skipped list:           none                     
=====> api docker options information
       Docker options build:                                   
       Docker options deploy:         --restart=on-failure:10 -v /var/lib/dokku/data/storage/ssl:/app/data 
       Docker options run:            -v /var/lib/dokku/data/storage/ssl:/app/data 
=====> api domains information
       Domains app enabled:           true                     
       Domains app vhosts:            api.mhoori.com           
       Domains global enabled:        true                     
       Domains global vhosts:         mhoori.com               
=====> api git information
       Git deploy branch:             master                   
       Git global deploy branch:      master                   
       Git keep git dir:              false                    
       Git rev env var:               GIT_REV                  
=====> api network information
       Network bind all interfaces:   false
       Network listeners:             172.17.0.12:5000
=====> api proxy information
       Proxy enabled:                 true                     
       Proxy type:                    nginx                    
       Proxy port map:                http:6001:6001 http:80:5000 https:443:5000
=====> api ps information
       Processes:                     9                        
       Deployed:                      true                     
       Running:                       true                     
       Restore:                       true                     
       Restart policy:                on-failure:10            
       Ps can scale:                  false                    
       Status queue.1:                running    (CID: 3df3c11113fd)
       Status queue.2:                running    (CID: 8db9d039a39d)
       Status queue.3:                running    (CID: 47fe44dd71ff)
       Status queue.4:                running    (CID: 4b954f2642aa)
       Status web.1:                  running    (CID: 015b31d9a32f)
       Status websocket.1:            running    (CID: 57f13dda05f0)
       Status websocket.2:            running    (CID: ecde67eec2c3)
       Status websocket.3:            running    (CID: 564dede7ae81)
       Status websocket.4:            running    (CID: 9e5bc7ea4762)
=====> api scheduler-docker-local information
       Scheduler docker local disable chown:                          
=====> api storage information
       Storage build mounts:                                   
       Storage deploy mounts:         -v /var/lib/dokku/data/storage/ssl:/app/data 
       Storage run mounts:            -v /var/lib/dokku/data/storage/ssl:/app/data
```

### How (deb/make/rpm) and where (AWS, VirtualBox, physical, etc.) was Dokku installed?:

Amazon AWS EC2

### Additional information

- Name of package: `beyondcode/laravel-websockets`

**Comment by josegonzalez** [comment]
> Scanning shows that port 6001 is not open

What does the word "scanning" mean? If your server has any firewall services, then you'll need to unblock the ports first. Common firewalls include:

- external: AWS Security Groups
- internal: ufw, iptables

**Comment by paschaldev** [comment]
@josegonzalez By scanning, I meant scanning for open PORTS on the server.

I use this PHP package https://github.com/beyondcode/laravel-websockets. 

On localhost, it starts the Websocket server on PORT 6001. When I deploy to Amazon EC2, from the logs, it starts too, I can see this with `dokku logs api`

```
2019-10-17T23:07:15.312729571Z app[websocket.2]: Starting the WebSocket server on port 6001...
```

However, when I scan for open port with `netcat` command with my private IP address inside my server, the only port opened is `5000` which is the buildpack port

```
Connection to 172.*.*.* 5000 port [tcp/*] succeeded!
...
nc: connect to 172.*.*.* port 6001 (tcp) failed: Connection refused
```

In my AWS Security group, port `6001` is open to public from anywhere, `ufw` is not enabled

**Comment by paschaldev** [comment]
My `nginx.conf` file

```
map $http_upgrade $type {
  default "ws";
  websocket "web";
}

server {
  listen      [::]:80;
  listen      80;
  server_name api.[mydomain].com; 
  access_log  /var/log/nginx/api-access.log;
  error_log   /var/log/nginx/api-error.log;

  return 301 https://$host:443$request_uri;

}

server {
  listen      [::]:443 ssl http2;
  listen      443 ssl http2;
  
  server_name api.[mydomain].com; 
  access_log  /var/log/nginx/api-access.log;
  error_log   /var/log/nginx/api-error.log;

  ssl_certificate           /home/dokku/api/tls/server.crt;
  ssl_certificate_key       /home/dokku/api/tls/server.key;
  ssl_protocols             TLSv1.2 TLSv1.3;
  ssl_prefer_server_ciphers off;

  keepalive_timeout   70;
  
  location / {
    try_files /nonexistent @$type;
  }

  location @ws {
    proxy_pass             http://api-websocket;
    proxy_set_header Host  $host;
    proxy_read_timeout     60;
    proxy_connect_timeout  60;
    proxy_redirect         off;

    # Allow the use of websockets
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection 'upgrade';
    proxy_set_header Host $host;
    proxy_cache_bypass $http_upgrade;
  }

  location    @web {

    gzip on;
    gzip_min_length  1100;
    gzip_buffers  4 32k;
    gzip_types    text/css text/javascript text/xml text/plain text/x-component application/javascript application/x-javascript application/json application/xml  application/rss+xml font/truetype application/x-font-ttf font/opentype application/vnd.ms-fontobject image/svg+xml;
    gzip_vary on;
    gzip_comp_level  6;

    proxy_pass  http://api-5000;
    http2_push_preload on; 
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $http_connection;
    proxy_set_header Host $http_host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header X-Forwarded-Port $server_port;
    proxy_set_header X-Request-Start $msec;
  }
  include /home/dokku/api/nginx.conf.d/*.conf;

  error_page 400 401 402 403 405 406 407 408 409 410 411 412 413 414 415 416 417 418 420 422 423 424 426 428 429 431 444 449 450 451 /400-error.html;
  location /400-error.html {
    root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
    internal;
  }

  error_page 404 /404-error.html;
  location /404-error.html {
    root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
    internal;
  }

  error_page 500 501 503 504 505 506 507 508 509 510 511 /500-error.html;
  location /500-error.html {
    root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
    internal;
  }

  error_page 502 /502-error.html;
  location /502-error.html {
    root /var/lib/dokku/data/nginx-vhosts/dokku-errors;
    internal;
  }
}

upstream api-5000 {

  server 172.*.*.*:5000;
}

upstream api-websocket {

  server 172.*.*.*:6001;
}
```

**Comment by josegonzalez** [comment]
Is the thing that is running the websocket server not in the `web` process?

**Comment by paschaldev** [comment]
No it's not. My profile has this 

````
websocket: php artisan websocket:serve --host=0.0.0 0 --port=6001
```

**Comment by josegonzalez** [comment]
Going to copy my answer from #3716 here:

> Ports are not attached to specific process types, but rather only to the web process type. This is by design - we follow the Heroku pattern here - and I would only attempt to fix this through a consulting agreement (though pull requests are always welcome).
>
> There is a related issue to inject listener variables for all processes in #3258, which might help with this issue, as then you could inject a custom nginx.conf.sigil to proxy a specific process, so I would follow along with that issue (or contribute a PR) if you are interested in this functionality.

Closing as this is otherwise a duplicate of #3258.

**Comment by paschaldev** [comment]
OK I swapped my build from a BUILDPACKS to `Dockerfile` just to see if I can acheive my goal. I already expose 2 ports, `9000` web and `6001` websocket.... Using the `proxy-ports:add` command I was able to bind `80:9000`. My web works fine.

Now from your comment above, would it work if I put the command in the `web` Process... 

#### Before
```
web: php artisan serve --port=9000 --host=0.0.0.0
queue: php artisan queue:work --sleep=3 --tries=1
release: php artisan migrate --force
websocket: php artisan websocket:serve --port=6001 --host=0.0.0.0
```


#### After
```
web: php artisan serve --port=9000 --host=0.0.0.0 &&  php artisan websocket:serve --port=6001 --host=0.0.0.0
queue: php artisan queue:work --sleep=3 --tries=1
release: php artisan migrate --force
```

Would the above work?

**Comment by josegonzalez** [comment]
No because the command after `&&` won't run until the serve command exits.

What you want to do - expose ports on two different process types - won't work in a single app until #3258 is completed. You'll need to deploy two different apps to do this.

**Comment by paschaldev** [comment]
Alright thanks @josegonzalez I'll deploy multiple apps and see if it works. Thank you for the help

**Comment by paschaldev** [comment]
@josegonzalez I was able to get Webocket & HTTP server running with `swooletw/laravel-swoole` which has HTTP & Websocket server implemented.

For anybody facing the same challenge, check out that package `swooletw/laravel-swoole`.

1. You'll have to change the way you use websocket as it doesn't use Laravel's broadcast but a direct implementation similar to `Socket.IO` and I also 
2. Make sure you bind to the correct port you set in the app
3. Use a `Dockerfile` deploy, so bye bye to buildpacks. This is my working `Dockerfile:

```
FROM php:7.2-fpm

# Set working directory
WORKDIR /app

# Install dependencies
RUN apt-get update && apt-get install -y \
    build-essential \
    libpng-dev \
    libjpeg62-turbo-dev \
    libfreetype6-dev \
    locales \
    zip \
    jpegoptim optipng pngquant gifsicle \
    vim \
    unzip \
    git \
    curl \
    libmagickwand-dev

# Clear cache
RUN apt-get clean && rm -rf /var/lib/apt/lists/*

# Install extensions
RUN docker-php-ext-install pdo_mysql mbstring zip exif pcntl
RUN docker-php-ext-configure gd --with-gd --with-freetype-dir=/usr/include/ --with-jpeg-dir=/usr/include/ --with-png-dir=/usr/include/
RUN docker-php-ext-install gd
RUN pecl install imagick
RUN pecl install swoole
RUN docker-php-ext-enable imagick swoole

# Install composer
RUN curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer

# Add user for laravel application
RUN groupadd -g 1000 www
RUN useradd -u 1000 -ms /bin/sh -g www www

# Copy existing application directory contents
COPY . /app

# Copy existing application directory permissions
# COPY --chown=www:www . /app

# Own the current directory
RUN chown www /app

# Change current user to www
USER www

# Install all PHP dependencies
RUN composer install --ignore-platform-reqs \
    --no-interaction \
    --no-plugins \
    --no-scripts \ 
    --no-suggest

# Copy files back to app root container
COPY . ./

USER root

ENTRYPOINT [""]

# Expose port
EXPOSE 1215

CMD [""]
```
4.  I used this in my `Procfile`
```
web: php artisan swoole:http start
release: php artisan migrate --force
```

After that I have an open HTTP and Websocket connection on the same URL `https://[domain.com]` and `wss://[domain].com` all work with socket

@josegonzalez Would you be kind enough to see the `Dockerfile` to make sure I'm not doing anything wrong. I would really appreciate.

**Comment by josegonzalez** [comment]
You probably don't need a custom dockerfile if your single process is going to expose both http and websocket, but this is probably fine as is.

**Comment by paschaldev** [comment]
Without the `Dockerfile`, it was unable to build properly with `heroku builpack php`, I was getting errors plus I needed to manually install `swoole` pecl extension. I tried installing with `predeploy` and was getting errors too.

I'll retry again without a dockerfile to see if it works. Thanks

**Comment by LeandroDeivis2020** [comment]
What is exactly @ws?, what make @?
